staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -95,6 +95,28 @@ type Reflector struct {
 	// etcd, which is significantly less efficient and may lead to serious performance and
 	// scalability problems.
 	WatchListPageSize int64
+	// Called whenever the ListAndWatch drops a connection, whether or not
+	// the connection had an error. So that clients can handle errors
+	// and implement backoff.
+	dropWatchHandler DropWatchHandler
+}
+
+type DropWatchHandler func(r *Reflector, err error, doneCh <-chan struct{})
```

Comment:
Why is "doneCh" passed here?

Reply:
this is kind of related to the above question: we want to make the backoff longer, or pause it entirely until manual user intervention.

But we also want to respect API guarantees around the case where the caller closes the done channel. does that make sense?

Reply:
Maybe--it definitely needs to be documented here what a good DropWatchHandler should do.

Did you consider letting people specify the retry frequency instead? We have a number of "backoff" mechanisms that are pluggable elsewhere. If all you want to do is control the frequency / retry behavior, that seems like a more natural intervention point.

Personally, I wouldn't have expected a handler like this to be blocking, I'd expect it to just be a notification.

Reply:
e.g., I think it'd be very reasonable for a "backoff" object to be passed to the reflector constructor.

Reply:
The problem I'm trying to solve isn't "Alice thinks the backoff should be 3 seconds, and Bob thinks the backoff should be 5 seconds, so let's make it an option!"

The problem is that this code isn't even the right layer of the API to handle errors. Is the end user looking at a terminal? Are they looking at a web dashboard? Or something other UI entirely? This code doesn't know where to send the error message or what additional context it needs.

Reply:
I 100% agree with that problem and this is close to being a fix--if you:
a) remove the channel
b) either add a comment explaining this is for error visibility and that one should offload any expensive processing OR launch this in a goroutine (I prefer the former)
c) Maybe change the name, ConnectionErrorHandler or WatchErrorHandler?

Then I can sign off on this.

(I think the other items from my last comment also deserve separate fixes)

Reply:
sounds good, I'll try to make this change later this afternoon

Reply:
@lavalamp all comments addressed! PTAL?

I'll follow up later with a PR for the backoff parameter

Reply:
> b) either add a comment explaining this is for error visibility and that one should offload any expensive processing OR launch this in a goroutine (I prefer the former)

Another approach is to attempt a non-blocking write of the error to a notification chan.